### ISSUE TITLE
Koreader 2020.09-35 PW4 crash CBZ file while zooming or progressing through the file

### ISSUE THREAD

**Issue by guidodavide** [koreader/koreader#6749]
* KOReader version: 2020.09-35
* Device: Kindle Paperwhite 4 5.13.2

#### Issue
Koreader crashes while rendering the next/prev page or zooming while reading CBZ files.
Kindle then displays the home screen. 50% of the crashes result in a Kindle UI freeze.
The only way to recover from it is force a reboot of the device.

#### Steps to reproduce
Open CBZ file, progress into the file, scroll up or down.
A crash can happen after several hours or even after a tens of minutes.
Typically it happens frequently after a device sleep/suspension.

##### `crash.log` (if applicable)
[crash.log](https://github.com/koreader/koreader/files/5323318/crash.log)

Crash was observed even with the stable KOReader 2020.09 "Peanut Butter".

KUAL is in use to launch Koreader, I tried both ASAP and standard launch options with no luck.
The CBZ file is always read with a 90 degree screen rotation (landscape mode), Panel zoom (manga/comic) is enabled.

Typically the crash happens when the next page/prev page is about to display (after touching the screen to progress) or when the Panel zoom is about to display (after holding the finger on the page to generate the panel zoom).
Basically when the touch input is received by Koreader.

Re-installed:
JailBreak-1.16.N-FW-5.x-hotfix
KUAL coplate - 44a61b6-20200604
kual-mrinstaller-1.7.N-r1782
few days ago, problem persists.

Can I enable even more logging to help diagnose the issue?

**Comment by NiLuJe** [comment]
Out Of Memory.

```
error: malloc of array (3056 x 11925 bytes) failed
./luajit: ./ffi/mupdf.lua:67: could not run page: malloc of array (3056 x 11925 bytes) failed (1)
```

That's ~35MB. Which doesn't look like much, but we're running in an *extremely* memory-constrained environment on Kindle.

Nothing much we can do about it, unless someone feels like reinventing the wheel to make running framework-less viable.

**Comment by Frenzie** [comment]
> That's ~35MB. Which doesn't look like much

No kidding. O_o

**Comment by ezdiy** [comment]
Won't fix, but there's a glimmer of hope - *some day* I plan on picviewer based on scanline (rather than whole-image) pipeline., specifically to deal with ginormous scans that routinely ask for 200M+ RAM now. Such a thing can still scale, crop, rotate and apply simple transforms like contrast, with no need for memory at all.

**Comment by Galunid** [comment]
@guidodavide You can still try to split this comic into smaller ones. Zooming shouldn't take too much memory, though there's definitely room for optimization. Actually I could reduce memory usage by ~50%, by freeing memory after variable outlived it's purpose, rather than doing that after function completed all the computations. I'm not sure whether it'd have any real impact though.

**Comment by guidodavide** [comment]
Thank you all.

I wasn't aware of this memory constraint. If it's only affecting the Kindle devices (or just some of them) shouldn't it be added to the wiki? (and possibly the workaround too).

It can avoid you some work on the open issues :)

**Comment by NiLuJe** [comment]
It didn't really used to be an actual issue outside of old devices and degenerate use cases, back in my days, but it apparently got oh so very much worse since.

**Comment by NiLuJe** [comment]
@ezdiy: On a vaguely related note, https://libvips.github.io/libvips/ looks fun :).

**Comment by ezdiy** [comment]
@NiLuJe Yup, vips is pretty much the idea (either use it directly, or just jury rig something like it at home). As long the input format is "sane" (image accepts zlib stream, output gives us scanlines), we should be able to display ridiculous RGBA 50Mpx scans.
